webbeans-junit5/src/main/java/org/apache/openwebbeans/junit5/internal/CdiExtension.java

```diff
@@ -172,4 +182,76 @@ private void doClose(final SeContainer container)
             }
         });
     }
+
+    private Cdi getCdiConfig(ExtensionContext extensionContext)
+    {
+        return AnnotationUtils.findAnnotation(extensionContext.getElement(), Cdi.class).orElse(null);
+    }
+
+    private SeContainer getContainer()
+    {
+        if (testInstanceContainer != null)
+        {
+            return testInstanceContainer;
+        }
+        else
+        {
+            return reusableContainer;
+        }
+    }
+
+    @Override
+    public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+            throws ParameterResolutionException
+    {
+        final Cdi config = extensionContext.getParent().map(this::getCdiConfig).orElse(null);
+        if (config == null || !config.injectParameters())
+        {
+            return false;
+        }
+
+        final SeContainer container = getContainer();
+        if (container == null)
+        {
+            return false;
+        }
+
+        Bean<?> bean = resolveParameterBean(container, parameterContext, extensionContext);
+        return bean != null;
+    }
+
+    @Override
+    public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+            throws ParameterResolutionException
+    {
+        final SeContainer container = getContainer();
+        if (container == null)
+        {
+            return false;
+        }
+
+        Bean<?> bean = resolveParameterBean(container, parameterContext, extensionContext);
+        BeanManager beanManager = container.getBeanManager();
+        CreationalContext<?> creationalContext = beanManager.createCreationalContext(bean);
+        creationalContexts.add(creationalContext);
+        return beanManager.getReference(bean, parameterContext.getParameter().getType(), creationalContext);
+    }
+
+    private Bean<?> resolveParameterBean(SeContainer container, ParameterContext parameterContext, ExtensionContext extensionContext)
+    {
+        BeanManager beanManager = container.getBeanManager();
+        Set<Bean<?>> beans = beanManager.getBeans(
+                parameterContext.getParameter().getType(),
+                getQualifiers(parameterContext.getParameter()));
+        return beanManager.resolve(beans);
```

Comment:
@rmannibucau suggested to consider caching the resolution results. I'm not yet sure exactly which would be the most promising approach or if it is really worth it or if the bean manager is optimized already and fast enough.

Reply:
I would just put a HashMap in the store of the extensioncontext and clear it in afterClass. Will save the resolution and array/qualifier setup (key is the type+annotations with a cached hashCode like https://github.com/apache/openwebbeans/blob/b3b9cb800da3d366b1d0df8111a530190cd3a0c7/webbeans-impl/src/main/java/org/apache/webbeans/container/BeanCacheKey.java#L77).

side note: you can also go to the bean instantation in supports since it will also check the context availability and if a context not active exception is thrown we should log it probably and ignore the bean - or fail, I'm not yet sure, wdyt?

Reply:
Could it be that `CdiExtension` as of now already calls
* org.apache.webbeans.container.BeanManagerImpl.getBeans(Type, Annotation...)
* org.apache.webbeans.container.InjectionResolver.implResolveByType(boolean, Type, Annotation...)
* org.apache.webbeans.container.InjectionResolver.implResolveByType(boolean, Type, Class<?>, Annotation...)
* org.apache.webbeans.container.BeanCacheKey.BeanCacheKey(boolean, Type, String, Function<Class<?>, AnnotatedType<?>>, Annotation...)

and caching it again would not speed the result up again as much?

Reply:
Well it would speed up in general and why you generally must not use Instance or Provider at runtime until needed but agree the instance must be cached more than the bean itself

Reply:
Regarding error handling you mention in "side note" I consider these cases relevant when a parameter is attempted to be resolved with a cdi bean:
* no bean found -> `supportsParameter` returns false and parameter is not resolved in this extension
* one bean found -> `supportsParameter` returns true and parameter is resolved in 
this extension
* more than one bean found / ambiguous -> `supportsParameter` raises an exception, currently an `AmbiguousResolutionException`

"found" above of course implies everything cdi usually consideres such as qualifiers, scopes, etc..

That can lead to a parameter not injected by cdi extension while a test expected that which will in an exception thrown by Junit and would most probably be caused and fixed in bean definitions, `@Cdi.classes`, etc.

All the spurious cases fail this way and nothing has to be logged or anything.

Reply:
> Well it would speed up in general and why you generally must not use Instance or Provider at runtime until needed but agree the instance must be cached more than the bean itself

I think that cdi already is able to resolve to the same bean instance if everything particularly scope fits. Is it what you meant?

Reply:
About "All the spurious cases fail this way and nothing has to be logged or anything.", ambiguous parameter injection does not throw AmbiguousResolutionException since it is not ambiguous for CDI but for junit and this one takes the first extensions returning true in supportsParameter. So question is: do we document how to solve that (custom qualifier making it not resolvable by cdi for ex) or create the skip API i thought about originally. Tempted to say both work for me after some thoughts, so I will let you decide ;).

About the speed up: factually it is not negligible so questions are:

1. Which does win: cache impl vs runtime overhead (tempted to say cache impl is trivial and its lifecycle trivial to handle so it is worth avoid twice the same instantiation)
2. What about @Dependent instances (should it be the same instance by test or a new one per injection point)
3. Do we care cause generally people will just inject most beans in the class (fields) and use parameter injections only for specific needs (maybe to get new instances as mentionned in 2).

So overall - and thanks to 3 - I tend to think your impl can be good enough to start (for next release) and we can gather feedback and refine it if needed.
However, it should be explicit in the doc how parameter injection is done, how cdi param injection can be skipped in favor of another junit extension and that field injection can be preferred for method stability (IDE support) + reuse accross instances (@testinstance). So likely a note in http://svn.apache.org/repos/asf/openwebbeans/cms-site/trunk/content/openwebbeans-junit5.mdtext would solve most of my concerns.

Reply:
In terms of instance yes but the overhead is negligible or not depending what you write. If you write a small "core" test suite with a reusable container you will slow down things.
That said if you take the option to split the parameter injections  in another extension then it is fully fine to impl it this way since you will not impact @Cdi + field injections users which is what we should avoid with this new feature IMHO.

Reply:
I guess your statement is wrong:

> ambiguous parameter injection does not throw AmbiguousResolutionException since it is not ambiguous for CDI but for junit and this one takes the first extensions returning true in supportsParameter

Ambiguous in cdi bean throws an AmbiguousResolutionException and more than one parameter resolver returning supportsParameter true makes JUnit throw ParameterResolutionException

> About the speed up: avoid twice the same instantiation

The current version of the code already avoids that. See a more recent version of the `ParameterResolutionTest` test near `testThatParameterGetsInjectedSameScopedBeanInstance`.

> That said if you take the option to split the parameter injections in another extension then it is fully fine to impl it this way since you will not impact @cdi + field injections users which is what we should avoid with this new feature IMHO.

I would not see how this PR would affect cdi field injection. It only adds a parameter resolver and does not (should not as far I'm aware of or it would be a bug). It could possibly make a test method parameter ambiguously resolved by more than one parameter resolver but then that would not affect field injection either would it?

Reply:
Ok, what I meant is that we already have @Cdi usages with several other jupiter extensions. This feature must not make working tests no more working, it would be a regression I don't see how to justify.
Currently the only way to handle it is @DontInject (which is ok).
This PR does not affect field injection but fully deactivating this new feature and using field injection as of today is a solution to this ambiguity issue.
Now, no issue to merge like that, we can rework it after.